src/Nonlinear/ReverseAD/utils.jl

```diff
@@ -60,6 +60,68 @@ function _UnsafeVectorView(x::Vector, N::Int)
     return _UnsafeVectorView(0, N, pointer(x))
 end
 
+"""
+    _UnsafeHessianView(x, N)
+
+Lightweight unsafe view that converts a vector `x` into the lower-triangular
+component of a symmetric `N`-by-`N` matrix.
+
+## Motivation
+
+`_UnsafeHessianView` is needed as an allocation-free equivalent of `view`. Other
+alternatives, like `reshape(view(x, 1:N^2), N, N)` or a struct like
+```julia
+struct _SafeView{T}
+    x::Vector{T}
+    len::Int
+end
+```
+will allocate so that `x` can be tracked by Julia's GC.
+`_UnsafeHessianView` relies on the fact that the use-cases of
+`_UnsafeHessianView` only temporarily wrap a long-lived vector like
+`d.jac_storage` so that we don't have to worry about the GC removing
+`d.jac_storage` while `_UnsafeHessianView` exists. This lets us use a `Ptr{T}`
+and create a struct that is `isbitstype` and therefore does not allocate.
+
+## Unsafe behavior
+
+`_UnsafeHessianView` is unsafe because it assumes that the vector `x` remains
+valid during the usage of `_UnsafeHessianView`.
+"""
+struct _UnsafeHessianView <: AbstractMatrix{Float64}
+    N::Int
+    ptr::Ptr{Float64}
+end
+
+Base.size(x::_UnsafeHessianView) = (x.N, x.N)
+
+function _linear_index(row, col)
+    if row < col
+        error("Unable to access upper-triangular component: ($row, $col)")
+    end
+    return div((row - 1) * row, 2) + col
+end
+
+function Base.getindex(x::_UnsafeHessianView, i, j)
+    return unsafe_load(x.ptr, _linear_index(i, j))
+end
+
+function Base.setindex!(x::_UnsafeHessianView, value, i, j)
+    unsafe_store!(x.ptr, value, _linear_index(i, j))
+    return value
+end
+
+function _UnsafeHessianView(x::Vector, N::Int)
+    z = div(N * (N + 1), 2)
+    if length(x) < z
+        resize!(x, z)
+    end
+    for i in 1:z
+        x[i] = 0.0
+    end
```

Comment:
How about using directly `fill!(x, 0.0)`?

Reply:
I don't think there is a difference? The loop is simpler because `fill!` actually calls a bunch of other methods like `size` etc.

Reply:
oh ok, that makes sense then!